swift/Arrow/Sources/Arrow/ArrowCExporter.swift

```diff
@@ -0,0 +1,133 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import Foundation
+import ArrowC
+import Atomics
+
+// The memory used by UnsafeAtomic is not automatically
+// reclaimed. Since this value is initialized once
+// and used until the program/app is closed it's
+// memory will be released on program/app exit
+let exportDataCounter: UnsafeAtomic<Int> = .create(0)
+
+public class ArrowCExporter {
+    private class ExportData {
+        let id: Int
+        init() {
+            id = exportDataCounter.loadThenWrappingIncrement(ordering: .relaxed)
+            ArrowCExporter.exportedData[id] = self
+        }
+    }
+
+    private class ExportSchema: ExportData {
+        public let arrowTypeName: UnsafePointer<CChar>
+        public let nameCstr: UnsafePointer<CChar>
+        private let arrowType: ArrowType
+        private let name: String
+        init(_ arrowType: ArrowType, name: String = "") throws {
+            self.arrowType = arrowType
+            // keeping the name str to ensure the cstring buffer remains valid
+            self.name = name
+            self.arrowTypeName = (try arrowType.cDataFormatId as NSString).utf8String!
+            self.nameCstr = (name as NSString).utf8String!
+            super.init()
+        }
+    }
+
+    private class ExportArray: ExportData {
+        private let arrowData: ArrowData
+        private(set) var data = [UnsafeRawPointer?]()
+        private(set) var buffers: UnsafeMutablePointer<UnsafeRawPointer?>
+        init(_ arrowData: ArrowData) {
+            // keep a reference to the ArrowData
+            // obj so the memory doesn't get
+            // deallocated
+            self.arrowData = arrowData
+            for arrowBuffer in arrowData.buffers {
+                data.append(arrowBuffer.rawPointer)
+            }
+
+            self.buffers = UnsafeMutablePointer(mutating: data)
+            super.init()
+        }
+    }
+
+    private static var exportedData = [Int: ExportData]()
```

Comment:
Is this container cleaned up at shutdown? I'm wondering what happens if a release callback is called after this container is cleaned up. Presumably the release callback would crash because of the `fatalError`?

This is not necessarily a problem, but I'm curious :-)

Reply:
Yes, the memory should be released during shutdown.  The container should be valid in the process that loaded the library until the process has been shutdown.

Reply:
Yes, my question was more along the lines of: what happens if the release callback is called _after_ finalization of Swift internals has started? Granted, we might not be able to do much about it, but it's probably good to keep this in mind (and/or add a comment about it).